### PR TITLE
Code-gen: Remove lowercasing for project imports

### LINKED ISSUE #20: Imports lowercased
Several packages are lowercased during package generation. This causes import errors if, for instance, your Github account uses uppercase characters.

Some packages are also generated under lowercased paths so these packages are generated under a different directory than your project. For example my project might be in $GOPATH/src/github.com/IanLewis/myapp and code-generator will generate some files under $GOPATH/src/github.com/ianlewis/myapp

### PR DESCRIPTION
**Pull request by jsturtevant**
**What this PR does / why we need it**:
If a project has a uppercase letter in the project import path then the generated paths will end up in different directories when some of the files are lower-cased during generation.  This PR fixes the few places where the imports are lower-cased.  

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/code-generator/issues/20

**Special notes for your reviewer**:
There was another PR (#63892) that was started but closed with no explanation.  It looks like it was missing a few spots area's where the lower casing was missed which are addressed in this PR.

I was not sure where or if there are tests for this section of code.  Would be happy to add tests if you can point me in the right direction.  

If you run the `code-generator/generate-groups.sh` against the project [here](https://github.com/Azure/azure-k8s-metrics-adapter/tree/code-gen-bug) you can see the lower-casing in action.  The generated code is split across two directories and the project does not build but with these fixes code-generation completes as expected and the project builds.

**Release note**:

```release-note
Code-gen: Remove lowercasing for project imports
```

### CODE REVIEW

**Comment by jsturtevant** [comment]
@caesarxuchao I was able to pass the tests with a few updates.  

I started to add an example of the Uppercasing in _examples like the previous PR.  The generation works as expected (I haven't commited this yet).

But during that further testing I notice that the verify tests don't actually run the smoke tests in the verify script.  From the [test output](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/68484/pull-kubernetes-verify/106590/?log#log) of this PR:

```
I0913 00:03:18.965] Smoke testing _example by compiling...
I0913 00:03:19.008] Building client-gen
W0913 00:03:19.108] warning: "vendor/k8s.io/code-generator/_example/..." matched no packages
W0913 00:03:19.109] warning: "vendor/k8s.io/code-generator/_example/..." matched no packages
```

There `_example` folder looks to have been renamed to _example**s** at some point.  I made an update to build the projects by updating `staging/src/k8s.io/code-generator/hack/verify-codegen.sh` to build the samples for example:

```
go build ${SCRIPT_ROOT}/_examples/crd/...
```

but now I get the error:

```
./hack/verify-codegen.sh
Generating deepcopy funcs
Generating defaulters
Generating conversions
Generating clientset for example:v1 example2:v1 at k8s.io/code-generator/_examples/apiserver/clientset
Generating listers for example:v1 example2:v1 at k8s.io/code-generator/_examples/apiserver/listers
Generating informers for example:v1 example2:v1 at k8s.io/code-generator/_examples/apiserver/informers
Generating deepcopy funcs
Generating clientset for example:v1 example2:v1 at k8s.io/code-generator/_examples/crd/clientset
Generating listers for example:v1 example2:v1 at k8s.io/code-generator/_examples/crd/listers
Generating informers for example:v1 example2:v1 at k8s.io/code-generator/_examples/crd/informers
Generating deepcopy funcs
Generating clientset for example:v1 at k8s.io/code-generator/_examples/UpperCaseProjectName/clientset
Generating listers for example:v1 at k8s.io/code-generator/_examples/UpperCaseProjectName/listers
Generating informers for example:v1 at k8s.io/code-generator/_examples/UpperCaseProjectName/informers
diffing ./hack/../_examples against freshly generated codegen
./hack/../_examples up to date.
Smoke testing _example by compiling...
./hack/..
../../../../vendor/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/clustertesttype.go:28:2: cannot find package "k8s.io/kubernetes/pkg/apis/autoscaling" in any of:
	/home/jstur/go/src/github.com/kubernetes/kubernetes/vendor/k8s.io/kubernetes/pkg/apis/autoscaling (vendor tree)
	/usr/local/go/src/k8s.io/kubernetes/pkg/apis/autoscaling (from $GOROOT)
	/home/jstur/go/src/k8s.io/kubernetes/pkg/apis/autoscaling (from $GOPATH)
```

Since the smoke tests have not been running it seems this has been an issue before this PR as you can see on this [file in master](https://github.com/kubernetes/kubernetes/blob/6eeff3e6c94452581a94d16769ea406dec848152/staging/src/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/clustertesttype.go#L28).  Looking at the vendor folder it seems that the all those packages are symlinked back into the main `pkg` folder.  Although none of them have the prefix `kubernetes`.  It seems that there might be another bug related to how those autoscaling imports are being generated.  

Thoughts on how to proceed?  I could check in the samples folder with the Uppercase as an example, get this merged and closed, then open a separate bug for the re-enabling the smoke tests and fixing the generating of the autoscaling package?  What do you think?

**Comment by jsturtevant** [comment]
@caesarxuchao all the tests pass is this good to merge?  I can open issues for the comments above separately.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/client_generator.go [hunk 1]
```diff
@@ -130,7 +130,7 @@ func DefaultNameSystem() string {
 }
 
 func packageForGroup(gv clientgentypes.GroupVersion, typeList []*types.Type, clientsetPackage string, groupPackageName string, groupGoName string, apiPath string, srcTreePath string, inputPackage string, boilerplate []byte) generator.Package {
-	groupVersionClientPackage := strings.ToLower(filepath.Join(clientsetPackage, "typed", groupPackageName, gv.Version.NonEmpty()))
+	groupVersionClientPackage := filepath.Join(clientsetPackage, "typed", groupPackageName, strings.ToLower(gv.Version.NonEmpty()))
```
Comment (caesarxuchao) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/client_generator.go]: Can you lower case the group package name as well?  Here and other places.

I don't see any reason why mixed case is needed there.

**Review by caesarxuchao — COMMENTED** [review]
Can you add a test demonstrating the client-gen works for camelcased path now?

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/client_generator.go (hunk 1 above)
Comment (jsturtevant) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/client_generator.go]: addressed in the latest updates.

**Comment by jsturtevant** [comment]
I added an example with MixedCase. Also needed to update the verify script to build the packages correctly as stated in my [previous comment](https://github.com/kubernetes/kubernetes/pull/68484#issuecomment-421143025).  

Once the example projects were building properly,  this uncovered a bug in the code generation for the sub resource `scale` in the existing crd example:

```
Smoke testing _example by compiling...
# k8s.io/kubernetes/vendor/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/fake
vendor/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/fake/fake_clustertesttype.go:137:46: not enough arguments in call to testing.NewRootGetSubresourceAction
	have (schema.GroupVersionResource, string)
	want (schema.GroupVersionResource, string, string)
```

I fixed this in the `generator_fake_for_type.go` by adding the correct parameters.  Everything builds nows.  I also tested the updates against [my larger project](https://github.com/Azure/azure-k8s-metrics-adapter/tree/code-gen-bug) that prompted this bug fix as mentioned in the initial PR.

**Comment by jsturtevant** [comment]
@caesarxuchao the latest commits passed the tests.  Ready for review.  Thanks!

**Comment by caesarxuchao** [comment]
LGTM. Thanks. Can you squash it to meaningful commits?

**Comment by jsturtevant** [comment]
@caesarxuchao squashed commits and tests passed.  Looks ready to go.  Thanks again!

**Comment by caesarxuchao** [comment]
/assign @lavalamp

for approval

File: staging/src/k8s.io/code-generator/hack/verify-codegen.sh [hunk 2]
```diff
@@ -50,4 +50,6 @@ fi
 
 # smoke test
 echo "Smoke testing _example by compiling..."
-go build ${SCRIPT_ROOT}/_example/...
\ No newline at end of file
+go build ./${SCRIPT_ROOT}/_examples/crd/...
+go build ./${SCRIPT_ROOT}/_examples/apiserver/...
+go build ./${SCRIPT_ROOT}/_examples/MixedCase/...
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/hack/verify-codegen.sh]: :+1:

**Pull request merged**